drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1181 @@
+		return -1;
+
+	/* set the CRTC object as active */
+	if (set_crtc_property(req, &out->crtc, out->crtc_id, "ACTIVE", 1) < 0)
+		return -1;
+
+	/* set properties of the plane related to the CRTC and the framebuffer */
+	if (set_plane_property(req, plane, out->plane_id, "FB_ID", buf->fb) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_ID", out->crtc_id) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "SRC_X", 0) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "SRC_Y", 0) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "SRC_W", buf->width << 16) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "SRC_H", buf->height << 16) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_X", 0) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_Y", 0) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_W", buf->width) < 0)
+		return -1;
+	if (set_plane_property(req, plane, out->plane_id, "CRTC_H", buf->height) < 0)
+		return -1;
+
+	return 0;
+}
+
+/*
+ * A short helper function to compute a changing color value. No need to
+ * understand it.
+ */
+
+static uint8_t next_color(bool *up, uint8_t cur, unsigned int mod)
+{
+	uint8_t next;
+
+	next = cur + (*up ? 1 : -1) * (rand() % mod);
+	if ((*up && next < cur) || (!*up && next > cur)) {
+		*up = !*up;
+		next = cur;
+	}
+
+	return next;
+}
+
+/*
+ * Draw on back framebuffer before the page-flip is requested.
+ */
+
+static void modeset_paint_framebuffer(struct modeset_output *out)
+{
+	struct modeset_buf *buf;
+	unsigned int j, k, off;
+
+	/* draw on back framebuffer */
+	out->r = next_color(&out->r_up, out->r, 5);
+	out->g = next_color(&out->g_up, out->g, 5);
+	out->b = next_color(&out->b_up, out->b, 5);
+	buf = &out->bufs[out->front_buf ^ 1];
+	for (j = 0; j < buf->height; ++j) {
+		for (k = 0; k < buf->width; ++k) {
+			off = buf->stride * j + k * 4;
+			*(uint32_t*)&buf->map[off] =
+				     (out->r << 16) | (out->g << 8) | out->b;
+		}
+	}
+
+}
+
+/*
+ * modeset_draw_out() prepares the framebuffer with the drawing and then it asks
+ * for the driver to perform an atomic commit. This will lead to a page-flip and
+ * the content of the framebuffer will be displayed. In this simple example
+ * we're only using the primary plane, but we could also be updating other
+ * planes in the same atomic commit.
+ *
+ * Just like in modeset_perform_modeset(), we first setup everything with
+ * modeset_atomic_prepare_commit() and then actually perform the atomic commit.
+ * The main difference is that here we have already painted the framebuffer and
+ * also we don't use the flag DRM_MODE_ALLOW_MODESET anymore, since the modeset
+ * already happened. If we continue using this flag, it may result in unecessary
+ * latency which is bad since we want to display content in a predictable and
+ * reasonable rate.
```

Comment:
> If we continue using this flag, it may result in unecessary latency which is bad since we want to display content in a predictable and reasonable rate.

Sort of yes, but not quite. There are two possible outcomes for an atomic commit:
- The atomic commit succeeds, and works exactly the same regardless of whether `ALLOW_MODESET` was set or not.
- If `ALLOW_MODESET` is set, the atomic commit succeeds but may take a while and cause monitor to go black for a moment. If `ALLOW_MODESET` is not set, the atomic commit fails.

The intention here is to always hit the first case. If we ever hit the second case, we have a bug in our code (or in the kernel), and in that case it's better to fail than glitch on. Someone else might argue that glitching on is better than failing.

The second case becomes really useful with `TEST_ONLY` commits, because if we use a `TEST_ONLY` commit then we probably have something else to fall back to rather that simply adding `ALLOW_MODESET` and retrying.

So, `ALLOW_MODESET` alone cannot cause anything. There must be something else in the KMS state that forces the DRM driver down the modeset path before `ALLOW_MODESET` makes a difference. If that something else is there, we can't pageflip to begin with.

Reply:
Thanks a lot for this explanation! I've edited the comment.